kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,328 @@
+            Integer such that x is subsampled by 2**k along the spatial variables.
+
+        Returns
+        -------
+        out : tensor_like
+            Tensor such that its Fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)].
+
+    """
+    def __call__(self, x, k):
+        batch_shape = x.shape[:-3]
+        signal_shape = x.shape[-3:]
+        x = x.view((-1,) + signal_shape)
+        y = x.view(-1,
+                       k, x.size(1) // k,
+                       k, x.size(2) // k,
+                       2)
+
+        out = y.mean(3, keepdim=False).mean(1, keepdim=False)
+        out = out.reshape(batch_shape + out.shape[-3:])
+        return out
+
+class Modulus(object):
+    """This class implements a modulus transform for complex numbers.
+
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+
+        Parameters
+        ---------
+        x : input tensor
+            Complex torch tensor.
+
+        Returns
+        -------
+        output : output tensor 
+            A tensor with the same dimensions as x, such that output[..., 0]
+            contains the complex modulus of x, while output[..., 1] = 0.
+
+    """
+    def __call__(self, x):
+        norm = torch.zeros_like(x)
+        norm[...,0] = (x[...,0]*x[...,0] +
+                       x[...,1]*x[...,1]).sqrt()
+        return norm
+
+def fft(x, direction='C2C', inverse=False):
+    """Interface with torch FFT routines for 2D signals.
+
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x)
+        x_ifft = fft(x, inverse=True)
+
+        Parameters
+        ----------
+        x : tensor
+            Complex input for the FFT.
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex.
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : If direction is equal to 'C2R', then an error is raised.
+        
+        Raises
+        ------
+        RuntimeError
+            In the event that we are going from complex to real and not doing
+            the inverse fft or in the event x is not contiguous.
+        TypeError
+            In the event that x does not have a final dimension 2 i.e. not
+            complex. 
+        
+        Returns
+        -------
+        output : tensor
+            Result of FFT or IFFT.
+            
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if not iscomplex(x):
+        raise TypeError('The input should be complex (e.g. last dimension is 2).')
+
+    if not x.is_contiguous():
+        raise RuntimeError('Tensors must be contiguous!')
+
+    if direction == 'C2R':
+        output = torch.irfft(x, 2, normalized=False, onesided=False) * x.size(-2) * x.size(-3)
```

Comment:
@edouardoyallon why is this multiplied by ` x.size(-2) * x.size(-3)`?

Reply:
@MuawizChaudhary This is an error reported by @nshervt and this will be fixed in a future PR.